lib/Dialect/RTL/Types.cpp

```diff
@@ -0,0 +1,62 @@
+//===- Types.cpp - RTL types code defs --------------------------*- C++ -*-===//
+//
+// Definitions for RTL data types. Anything which doesn't have to be public
+// should go in here.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/RTL/Types.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/DialectImplementation.h"
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/Hashing.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+using namespace mlir;
+using namespace circt::rtl;
+
+namespace circt {
+namespace rtl {
+
+/// Equality operator for FieldInfo. Required for data types in TypeStorage
+static bool operator==(const FieldInfo &a, const FieldInfo &b) {
+  return a.name == b.name && a.type == b.type;
+}
+
+/// Compute a hash for FieldInfo. Required for data types in TypeStorage
+static llvm::hash_code hash_value(const FieldInfo &fi) { // NOLINT
+  return llvm::hash_combine(fi.name, fi.type);
+}
+
+} // namespace rtl
+} // namespace circt
+
+#define GET_TYPEDEF_CLASSES
+#include "circt/Dialect/RTL/RTLTypes.cpp.inc"
+
+Type EnumType::parse(MLIRContext *ctxt, DialectAsmParser &parser) {
```

Comment:
Just curious, what is the reason to define the parser/printer here rather than in ODS like the other types?

Reply:
The ODS spec is because it has to be in order for StructType and UnionType to inherit it. This is here since it can be and I generally prefer not writing c++ in tblgen.

Reply:
On second thought, I'll pull these out of here. They're a lot of code to put in ODS, IMO.